include/i2s.h

```diff
@@ -0,0 +1,418 @@
+	/** @brief The interface is receiving / transmitting data. */
+	I2S_STATE_RUNNING,
+	/** @brief The interface is flushing its transmit queue. */
+	I2S_STATE_STOPPING,
+	/** @brief Transmit buffer underrun has occurred. */
+	I2S_STATE_UNDERRUN,
+	/** @brief Receive buffer overrun has occurred. */
+	I2S_STATE_OVERRUN,
+};
+
+/** Trigger command */
+enum i2s_trigger_cmd {
+	/** @brief Start the transmission / reception of data.
+	 *
+	 * The command can only be issued after the interface was configured.
+	 * Additionally if I2S_DIR_TX is set some data has to be queued for
+	 * transmission by the i2s_write() function. This trigger moves the
+	 * interface to the RUNNING state.
+	 */
+	I2S_TRIGGER_START,
+	/** @brief Stop the transmission / reception of data.
+	 *
+	 * Stop the transmission / reception of data at the end of the current
+	 * memory block. This trigger moves the interface to the READY state.
+	 * Subsequent START trigger will resume transmission / reception where
+	 * it stopped.
+	 */
+	I2S_TRIGGER_STOP,
+	/** @brief Flush the transmit queue.
+	 *
+	 * Send all data in the transmit queue and stop the transmission /
+	 * reception. This trigger moves the interface to the READY state.
+	 */
+	I2S_TRIGGER_FLUSH,
+	/** @brief Discard the transmit queue.
+	 *
+	 * Stop the transmission / reception immediately and discard the
+	 * contents of the respective queue. This trigger moves the interface to
+	 * the READY state.
+	 */
+	I2S_TRIGGER_DROP,
+	/** @brief Prepare the queues after underrun/overrun error has occurred.
+	 *
+	 * When underrun/overrun error occurs this trigger will prepare the
+	 * queues and move the interface to the READY state.
+	 */
+	I2S_TRIGGER_PREPARE,
+};
+
+/** @struct i2s_config
+ * @brief Interface configuration options.
+ *
+ * Memory slab pointed to by the mem_slab field has to be defined and
+ * initialized by the user. For I2S driver to function correctly number of
+ * memory blocks in a slab has to be at least 2. Please check Zephyr Kernel
+ * Primer for more information on memory slabs.
+ *
+ * @remark In I2S mode parameter channels is ignored, number of words in frame
+ * is always 2.
+ *
+ * @param word_size Number of bits representing one data word.
+ * @param channels Number of channels, this is also number of words per frame.
+ * @param format Data stream format as defined by I2S_FMT_* constants.
+ * @param options Configuration options as defined by I2S_OPT_* constants.
+ * @param frame_clk_freq Frame clock (WS) frequency.
+ * @param mem_slab memory slab to store RX/TX data.
+ * @param block_size Size of one RX/TX memory block (buffer) in bytes.
+ * @param timeout Read/Write timeout. Number of milliseconds to wait in case TX
+ *        queue is full, RX queue is empty or one of the special values
+ *        K_NO_WAIT, K_FOREVER.
+ */
+struct i2s_config {
+	u8_t word_size;
+	u8_t channels;
+	i2s_fmt_t format;
+	i2s_opt_t options;
+	u32_t frame_clk_freq;
+	struct k_mem_slab *mem_slab;
```

Comment:
mem_slab here is a pointer to the pool of TX/RX memory blocks, not a single buffer. That's how I2S driver knows where to release TX memory blocks and where from reserve RX memory blocks. Unlike in networking API memory slabs belong to/are initialized by the application, not the driver.

I'll update inline documentation to provide more information about initializing memory slabs.